File: Core/Features/Products/CreateProduct.cs
namespace Core.Features.Products;

public static class CreateProduct
{
    public record Command(
        int TenantCode,
        string Name,
        string? Description,
        decimal Price,
        IEnumerable<string> TagCodes) : IRequest;

    public class Handler(
        ITenantRepository tenantRepository,
        ITagRepository tagRepository,
        IProductRepository productRepository) : IRequestHandler<Command>
    {
        public async Task Handle(Command request, CancellationToken cancellationToken)
        {
            if (!await tenantRepository.ExistsByCodeAsync(request.TenantCode, cancellationToken))
            {
                throw new TenantNotFoundException(request.TenantCode);
            }

            var existingTagCodes = await tagRepository.ExistsByCodeAsync(request.TagCodes, cancellationToken);
            if (!existingTagCodes)
            {
                throw new TagNotFoundException(request.TagCodes);
            }

            var product = new Product(
                request.TenantCode,
                request.Name,
                request.Description,
                request.Price,
                request.TagCodes);

            await productRepository.CreateAsync(product, cancellationToken);
        }
    }
}

File: backoffice/src/server/queries/products/mutations/create-product.ts
import { fetchGraphQL } from '@/utils'

export async function createProductMutation(
  tenantCode: number,
  name: string,
  description: string,
  price: number,
  tagCodes: string[]
): Promise<string> {
  const mutation = `
    mutation CreateProduct($tenantCode: Int!, $name: String!, $description: String!, $price: Decimal!, $tagCodes: [String!]!) {
      createProduct(command: {
        tenantCode: $tenantCode,
        name: $name,
        description: $description,
        price: $price,
        tagCodes: $tagCodes,
      }) {
        code
      }
    }
  `

  const result = await fetchGraphQL<{ createProduct: { code: string } }>(
    mutation,
    {
      tenantCode,
      name,
      description,
      price,
      tagCodes,
    }
  )

  return result.createProduct.code
}
